<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [6]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [7]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
vacancies = f'''select * 
            from vacancies 
            '''
df = pd.read_sql_query(vacancies, connection)
df.info()

C:\Users\ciclone\AppData\Local\Temp\ipykernel_18852\3120257653.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(vacancies, connection)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49197 entries, 0 to 49196
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           49197 non-null  int64  
 1   name         49197 non-null  object 
 2   key_skills   40741 non-null  object 
 3   schedule     49197 non-null  object 
 4   experience   49197 non-null  object 
 5   employment   49197 non-null  object 
 6   salary_from  21485 non-null  float64
 7   salary_to    12690 non-null  float64
 8   area_id      49197 non-null  int64  
 9   employer_id  49197 non-null  int64  
dtypes: float64(2), int64(3), object(5)
memory usage: 3.8+ MB


### Количество вакансий в базе 49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [5]:
# текст запроса
employers = f'''select * 
            from employers 
            '''
df = pd.read_sql_query(employers, connection)
df.info()

C:\Users\ciclone\AppData\Local\Temp\ipykernel_18852\3182452119.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(employers, connection)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23501 entries, 0 to 23500
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      23501 non-null  int64 
 1   name    23501 non-null  object
 2   area    23501 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 550.9+ KB


### Количество работодателей в базе 23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [6]:
# текст запроса
areas = f'''select * 
            from areas 
            '''
df = pd.read_sql_query(areas, connection)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1362 entries, 0 to 1361
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      1362 non-null   int64 
 1   name    1362 non-null   object
dtypes: int64(1), object(1)
memory usage: 21.4+ KB


C:\Users\ciclone\AppData\Local\Temp\ipykernel_18852\1939134718.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(areas, connection)


### Количество регионов в базе 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [7]:
# текст запроса
industries = f'''select * 
            from industries 
            '''
df = pd.read_sql_query(industries, connection)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      294 non-null    object
 1   name    294 non-null    object
dtypes: object(2)
memory usage: 4.7+ KB


C:\Users\ciclone\AppData\Local\Temp\ipykernel_18852\2621171229.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(industries, connection)


### Количество сфер деятельности 294

***

### Для работы нам представлена база данных из 4-х основных таблиц и одной дополнительной, а именно: 
### вакансии (VACANCIES) количестко записей: 49197 
### работодатели (EMPLOYERS) количестко записей: 23501
### регион (AREAS) количестко записей: 1362
### сферы деятельности (INDUSTRIES) количестко записей: 294


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
# текст запроса
cnt_area = connection.cursor()
cnt_area.execute("SELECT areas.name, COUNT(vacancies.area_id) AS cnt FROM vacancies JOIN areas ON vacancies.area_id = areas.id GROUP BY areas.id ORDER BY cnt DESC")

rows = cnt_area.fetchall()

for row in rows:
    print(row)

cnt_area.close()
connection.close()
    

#### пятёрка лидеров по количеству вакансий:
#### Москва, Санкт-Петербург, Минск, Новосибирск, Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
cnt_salafy_from_to = connection.cursor()
cnt_salafy_from_to.execute("SELECT COUNT(*) FROM vacancies WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL")

rows = cnt_salafy_from_to.fetchall()

for row in rows:
    print(row)

cnt_salafy_from_to.close()
connection.close()

(24073,)


#### Вакинсий по такому запросу 24 073 

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [18]:
avg_salafy_from_to = connection.cursor()
avg_salafy_from_to.execute("SELECT ROUND(AVG(salary_from)) as avg_salary_from, ROUND(AVG(salary_to)) as avg_salary_to FROM vacancies")

rows = avg_salafy_from_to.fetchall()

for row in rows:
    print(row)

avg_salafy_from_to.close()



(Decimal('71065'), Decimal('110537'))


#### Средняя минимальная зарплата 71065, средняя максимальная зарплата 110537

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
sched_empl = connection.cursor()
sched_empl.execute("SELECT schedule, employment, COUNT(*) as count FROM vacancies GROUP BY schedule, employment ORDER BY count DESC")

rows = sched_empl.fetchall()

for row in rows:
    print(row)

sched_empl.close()

#### Вторая по популярности пара: Удаленная работа, полная занятость

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
exp_sort = connection.cursor()
exp_sort.execute("SELECT experience, COUNT(*) as count FROM vacancies GROUP BY experience ORDER BY count ASC")

rows = exp_sort.fetchall()

for row in rows:
    print(row)

exp_sort.close()

('Более 6 лет', 1337)
('Нет опыта', 7197)
('От 3 до 6 лет', 14511)
('От 1 года до 3 лет', 26152)


#### Более 6 лет — Нет опыта — От 3 до 6 лет — От 1 года до 3 лет

***

#### Из всех вакансий, почти половина указывает уровень желаемого дохода. Основное количество вакансий из крупнейших городов своих регионов, более половины из которых с опытом работы от 1 года до 3 лет

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:

empl_1_5 = connection.cursor()
empl_1_5.execute("SELECT employers.name, COUNT(vacancies.id) AS vacancies_count FROM employers JOIN vacancies ON employers.id = vacancies.employer_id GROUP BY employers.name ORDER BY vacancies_count DESC limit 5")

rows = empl_1_5.fetchall()

for row in rows:
    print(row)

empl_1_5.close()

('Яндекс', 1933)
('Ростелеком', 491)
('Тинькофф', 444)
('СБЕР', 428)
('Газпром нефть', 331)


#### Яндекс — Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [10]:

empl_zero_vac = connection.cursor()
empl_zero_vac.execute("SELECT areas.name, COUNT(DISTINCT employers.id) AS employers_count FROM areas left JOIN employers ON employers.area = areas.id LEFT JOIN vacancies ON vacancies.area_id = employers.area WHERE vacancies.id is null GROUP BY areas.name order by employers_count desc limit 1")

rows = empl_zero_vac.fetchall()

for row in rows:
    print(row)

empl_zero_vac.close()

('Россия', 410)


#### По всей видимости мы наблюдаем ошибки при заполнении данных, где указана только страна и нет конкретного региона

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [9]:

empl_reg_max = connection.cursor()
empl_reg_max.execute("SELECT MAX(regions_count) FROM (SELECT employers.id, COUNT(DISTINCT areas.id) AS regions_count FROM employers JOIN vacancies ON employers.id = vacancies.employer_id JOIN areas ON vacancies.area_id = areas.id GROUP BY employers.id) AS regions_per_employer")

rows = empl_reg_max.fetchall()

for row in rows:
    print(row)

empl_reg_max.close()

(181,)


#### Среди работодателей существует компания которая размещает свои вакансии в 181 регионах

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [8]:
empl_no_inds = connection.cursor()
empl_no_inds.execute("select count(*) cnt from employers e left join employers_industries ei on e.id = ei.employer_id where ei.industry_id is null")

rows = empl_no_inds.fetchall()

for row in rows:
    print(row)

empl_no_inds.close()

(8419,)


#### 8419 Компаний не указали сферу деятельности

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [9]:
empl_4_3 = connection.cursor()
empl_4_3.execute("SELECT e.name, COUNT(ei.industry_id) AS industry_count FROM employers e JOIN EMPLOYERS_INDUSTRIES ei ON e.id = ei.employer_id group BY e.id HAVING COUNT(ei.industry_id) = 4 order by e.name offset 2 limit 1 ")

rows = empl_4_3.fetchall()

for row in rows:
    print(row)

empl_4_3.close()

('2ГИС', 4)


#### 2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [10]:
ind_it = connection.cursor()
ind_it.execute("SELECT COUNT(*) FROM employers e JOIN EMPLOYERS_INDUSTRIES ei ON e.id = ei.employer_id WHERE ei.industry_id = (SELECT id FROM industries WHERE name = 'Разработка программного обеспечения')")

rows = ind_it.fetchall()

for row in rows:
    print(row)

ind_it.close()

(3553,)


#### 3553 Компании из сферы разработки программного обеспечения 

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [27]:
# код для получения списка городов-милионников
import urllib.parse as req 
pathcommon = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионерыРоссии#Список_городов-миллионеров') 
url = path_common.scheme + '://' + req.quote(path_common.netloc) + req.quote(path_common.path) + '#'+ req.quote(path_common.fragment) 
url
df = pd.read_html(url)[1]

cities = tuple(df['Город'])
column_list = list(cities)
column_list = [x.replace('[a]', '') for x in column_list]
cities = tuple(column_list)
display(cities)


('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [32]:
# текст запроса

import sqlalchemy
from sqlalchemy import text
import pandas as pd

# установка соединения
url = f'postgresql+psycopg2://skillfactory:cCkxxLVrDE8EbvjueeMedPKt@84.201.134.129:5432/project_sql'
engine = sqlalchemy.create_engine(url)
conn = engine.connect()

# ЗАПРОС
query = '''
    select a.name,
       count(*) cnt
from employers e
         join vacancies v on e.id = v.employer_id
         join areas a on a.id = v.area_id
where e.name = 'Яндекс'
      and a.name in {cities}
group by a.id

union all

select 'Total', count(*)
from vacancies v
     join employers e on e.id = v.employer_id
     join areas a on a.id = v.area_id
where e.name = 'Яндекс'
      and a.name in {cities} 
order by cnt
    '''


#### Строк в выборке 17, Total 485

***

#### Количество вакансий в базе данных составляет 49197.
#### Количество работодателей в базе данных составляет 23501.
#### Количество регионов в базе данных составляет 1362.
#### Количество сфер деятельности в базе данных составляет 294.
#### Пять регионов с наибольшим количеством вакансий: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы.
#### Количество вакансий, у которых заполнено хотя бы одно из двух полей с зарплатой, составляет 24073.
#### Средняя минимальная зарплата составляет 71065, а средняя максимальная зарплата - 110537.
#### Работодатели, занимающие первое и пятое места по количеству вакансий: "Яндекс" (1933), "Ростелеком" (491), "Тинькофф" (444), "СБЕР" (428), #### "Газпром нефть" (331).
#### Количество компаний, которые не указали сферу деятельности, составляет 8419.
#### Количество компаний из сферы разработки программного обеспечения составляет 3553

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [4]:
data_vac = connection.cursor()
data_vac.execute("SELECT COUNT(*) FROM vacancies WHERE LOWER(name) LIKE '%данн%' OR LOWER(name) LIKE '%data%'")

rows = data_vac.fetchall()

for row in rows:
    print(row)

data_vac.close()

(1771,)


#### 1771 вакансия имеет отношение к данным

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [5]:
it_vac = connection.cursor()
it_vac.execute("select count(*) from vacancies v where (v.name ~* 'data scien(ce|tist)|исследователь данных|ML|machine learning|машинн.*обучен.*' and v.name !~* 'HTML') and (v.name ~* 'junior' or v.experience = 'Нет опыта' or v.employment = 'Стажировка')")

rows = it_vac.fetchall()

for row in rows:
    print(row)

it_vac.close()

(51,)


#### 51 вакансиямия для дата-сайентистов

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [4]:
count_sql_pstgrs = connection.cursor()
count_sql_pstgrs.execute("SELECT COUNT(*) FROM vacancies v WHERE (LOWER(v.name) LIKE '%data scientist%' OR LOWER(v.name) LIKE '%data science%' OR LOWER(v.name) LIKE '%исследователь данных%' OR v.name LIKE '%ML%' AND LOWER(v.name) NOT LIKE '%html%' OR LOWER(v.name) LIKE '%machine learning%' OR LOWER(v.name) LIKE '%машинн%обучен%') AND (LOWER(v.key_skills) LIKE '%sql%' OR LOWER(v.key_skills) LIKE '%postgres%')")

rows = count_sql_pstgrs.fetchall()

for row in rows:
    print(row)

count_sql_pstgrs.close()

(201,)


#### 201 Вакансия DS в ключевых навыках SQL или postgres

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [12]:
count_python = connection.cursor()
count_python.execute("SELECT COUNT(*) FROM vacancies v WHERE (LOWER(v.name) LIKE '%data scientist%' OR LOWER(v.name) LIKE '%data science%' OR LOWER(v.name) LIKE '%исследователь данных%' OR v.name LIKE '%ML%' AND LOWER(v.name) NOT LIKE '%html%' OR LOWER(v.name) LIKE '%machine learning%' OR LOWER(v.name) LIKE '%машинн%обучен%') AND (LOWER(v.key_skills) LIKE '%python%')")

rows = count_python.fetchall()

for row in rows:
    print(row)

count_python.close()

(351,)


#### 351 Вакансия DS в ключевых навыках указан Python

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
count_key_skill = connection.cursor()
count_key_skill.execute("SELECT AVG(num_skills) FROM (SELECT v.name, v.key_skills, (LENGTH(key_skills) - LENGTH(REPLACE(key_skills, ',', '')) + 1) AS num_skills FROM vacancies v WHERE LOWER(v.name) LIKE '%data scientist%' OR LOWER(v.name) LIKE '%data science%' OR LOWER(v.name) LIKE '%исследователь данных%' OR v.name LIKE '%ML%' AND LOWER(v.name) NOT LIKE '%html%' OR LOWER(v.name) LIKE '%machine learning%' OR LOWER(v.name) LIKE '%машинн%обучен%') subq")

rows = count_key_skill.fetchall()

for row in rows:
    print(row)

count_key_skill.close()

#### 6.41 в среднем ключевых навыков в вакансиях DS

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
solary_DS = connection.cursor()
solary_DS.execute("SELECT ROUND(AVG((COALESCE(v.salary_from, 0) + COALESCE(v.salary_to, 0)) / 2)) FROM vacancies v WHERE (LOWER(v.name) LIKE '%data scientist%' OR LOWER(v.name) LIKE '%data science%' OR LOWER(v.name) LIKE '%исследователь данных%' OR v.name LIKE '%ML%' OR LOWER(v.name) LIKE '%machine learning%' OR LOWER(v.name) LIKE '%машинн%обучен%') AND LOWER(v.name) NOT LIKE '%html%' AND experience = 'От 3 до 6 лет' AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)")

rows = solary_DS.fetchall()

for row in rows:
    print(row)

solary_DS.close()


#### 172245

***

In [ ]:
#### 

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

In [8]:
top_10_salary = connection.cursor()
top_10_salary.execute("SELECT v.name AS vacancy_name, e.name AS employer_name, (COALESCE(v.salary_from, 0) + COALESCE(v.salary_to, 0)) / 2 AS avg_salary FROM vacancies v JOIN employers e ON v.employer_id = e.id ORDER BY avg_salary DESC, employer_name LIMIT 10")

rows = top_10_salary.fetchall()

for row in rows:
    print(row)

top_10_salary.close()

('Брокер по продаже апартаментов', 'Агентство недвижимости ЛЕТО', 750000)
('СТО/Директор по разработке', 'Шеф Маркет', 750000)
('FullStack Developer in Unicorn Company', 'Петухова Алия', 717000)
('Frontend Developer in Unicorn Company', 'Петухова Алия', 717000)
('QA Automation Engineer', 'iBEC Systems', 700000)
('1С:Эксперт по производительности 1С', 'ЦТП', 650000)
('Product manager LATOKEN', 'LATOKEN', 600000)
('Product manager LATOKEN', 'LATOKEN', 600000)
('Senior DevOps in Blockchain Startup', 'Петухова Алия', 587500)
('Senior Software Developer (upwork / иностранный заказчик)', 'RuWork', 555261)


#### Если обратить внимание на Топ-10 самых высокооплачиваемых вакансий по всей базе, то интересной тенденцией является факт, что 9 из 10 вакансий из IT-сферы, что говорит о возросшем спросе и готовности со стороны работодателей достойно оплачивать специалистов данной сферы, и рынок труда в этом направлении продолжает расти.